dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,313 @@
+   */
+  void PopPixel() {
+    assert(elements_ > 0);
+    elements_--;
+    start_++;
+    WrapPosition(start_);
+  }
+
+  /**
+   * @brief Add a pixel, represented by contiguous `NumChannels()` elements and starting at `input`,
+   * to the buffer.
+   */
+  void PushPixel(const T* input) {
+    assert(elements_ < length_);
+    for (int c = 0; c < NumChannels(); c++) {
+      data_[end_ * NumChannels() + c] = *input;
+      input++;
+    }
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Add a pixel to the buffer
+   */
+  void PushPixel(span<const T> input) {
+    assert(elements_ < length_);
+    for (int c = 0; c < NumChannels(); c++) {
+      data_[end_ * NumChannels() + c] = input[c];
+    }
+    elements_++;
+    end_++;
+    WrapPosition(end_);
+  }
+
+  /**
+   * @brief Get pointer to the start of pixel `idx` in the buffer, taking into account the cyclic
+   * wrapping
+   */
+  T* GetPixelOffset(int idx) {
+    assert(idx < elements_);
+    if (start_ + idx < length_) {
+      return data_ + (start_ + idx) * NumChannels();
+    } else {
+      return data_ + (start_ + idx - length_) * NumChannels();
+    }
+  }
+
+  /**
+   * @brief Calculate dot product with 1-channel `window` length equal to the one specified
+   * at construction. The result is stored in pixel stored at `accum`.
+   */
+  template <typename W>
+  void CalculateDot(W* accum, const W* window) {
+    assert(elements_ == length_);
+    for (int c = 0; c < NumChannels(); c++) {
+      accum[c] = 0;
+    }
+    int window_idx = 0;
+    for (int buf_idx = start_; buf_idx < length_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumChannels(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumChannels() + c];
+      }
+    }
+    for (int buf_idx = 0; buf_idx < end_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumChannels(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumChannels() + c];
+      }
+    }
+  }
+
+  int Size() {
+    return elements_;
+  }
+
+  bool Empty() {
+    return elements_ == 0;
+  }
+
+ private:
+  void WrapPosition(int& pos) {
+    if (pos == length_) {
+      pos = 0;
+    }
+  }
+
+  template <bool has_channels_ = has_channels>
+  std::enable_if_t<has_channels_, int> NumChannels() {
+    return num_channels_;
+  }
+
+  template <bool has_channels_ = has_channels>
+  std::enable_if_t<!has_channels_, int> NumChannels() {
+    return 1;
+  }
+
+  T* data_ = nullptr;
+  int start_ = 0;
+  int end_ = 0;  ///< next empty element
+  int elements_ = 0;
+  int length_ = 0;
+  int num_channels_ = 0;
+};
+
+template <typename T, bool has_channels>
+void load_pixel_with_border(CyclicPixelWrapper<T, has_channels>& cpw, const T* in_ptr, int in_idx,
+                            int stride, int axis_size, span<const T> fill_value) {
+  cpw.PushPixel(in_ptr + boundary::idx_reflect_101(in_idx, axis_size) * stride);
```

Comment:
Can you make the border policy a parameter?

Reply:
Can we make it in separate PR? I don't think we support any enum that would indicate the border policy for any operator, I guess adding that separately would be just a trivial portion of boilerplate.

Reply:
Sure.